### PR TITLE
Show all annotations as top-level cards in streamer

### LINKED ISSUE #1916: Too many message not available on stream
When I open a stream page I often see too many "message not available" cards with some replies.
This is very confusing for a normal user and he/she may believe that the stream is not working.
bad_stream
 
The cause of this error that our search requests are sorted (and sadly this is "hardcoded" in the backend) by the updated field descending.

This means that the replies (as they were created later then the parent) are retrieved before their parents thus resulting in "Message not available" cards with replies that you can open. 
Of course scrolling down solves this as by the inf. scrolling feature the parents will be loaded eventually.

We have to seriously improve this.

### PR DESCRIPTION
**Pull request by gergely-ujvari**
First step for streamer and viewer separation.
StreamSearchController generates the cards to be displayed
in the streamer, and handles the annotation events.

Stream no longer relies on the threading service.
The messageContainers from jwz are still used.
They still share the same template.

Replies in the stream are displayed as root level annotation
cards.

Fix #1916

### CODE REVIEW

**Comment by gergely-ujvari** [comment]
Third version of #2061 
Second version of #2069 

The cards class is not a service, because we do not use it anywhere else

**Comment by tilgovi** [comment]
It's fine, but a little OO for my taste.

Can we keep it super simple and just inline the callbacks in the controller? The `cards` root container never changes, it can just be a local variable.

**Comment by gergely-ujvari** [comment]
Okay, deOOPed, handled the login/logout case and added tests for StreamSearchController.

Ready for review.

For the further separation I'll open a separate issue in our vision repo

**Comment by gergely-ujvari** [comment]
Actually I opened https://github.com/hypothesis/h/issues/2089 for it, and that can be linked to https://github.com/hypothesis/vision/issues/166

**Comment by gergely-ujvari** [comment]
Actually it needs a bit more work.

**Comment by gergely-ujvari** [comment]
Okay. Now this is really done.
I moved the threading service from AppController, added hook functions for looking up annotations, updated the tests.

**Comment by nickstenning** [comment]
Rebased this for you.

File: h/static/scripts/test/annotation-viewer-controller-test.coffee [hunk 1]
```diff
@@ -5,17 +5,81 @@ sinon.assert.expose assert, prefix: null
 
 
 describe 'AnnotationViewerController', ->
-  annotationViewerController = null
+  $scope = null
```
Comment (nickstenning) [inline h/static/scripts/test/annotation-viewer-controller-test.coffee]: Duplicate line.

File: h/static/scripts/test/streamsearch-test.coffee [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+
+assert = chai.assert
+sinon.assert.expose assert, prefix: null
+
+describe 'h:streamsearch', ->
+  before ->
+    angular.module('h', ['ngRoute'])
+    require('../streamsearch')
+
+  describe 'StreamSearchController', ->
```
Comment (nickstenning) [inline h/static/scripts/test/streamsearch-test.coffee]: I think this must be my fault... going to look into what happened to the rebase here.

**Comment by gergely-ujvari** [comment]
@nickstenning: Thank you very much for this!

**Comment by gergely-ujvari** [comment]
Can I get a reviewer for this PR?

**Comment by tilgovi** [comment]
The only general uncertainty I have about this is that we're using the scope to register different implementations depending on our controller. We know that a separation of the templates might be a good idea, but part of that is really about making the `AppController` also not responsible for too much. I wonder if we want to move the `streamer.onmessage` assignment inside the view controllers and just let each view handle updates in its own way.

**Comment by gergely-ujvari** [comment]
@tilgovi: What about the `cleanupAnnotations()` inside the AppController? That is responsible for deleting annotations clientside.
Should any template have this power? (If yes, than they cannot really coexists)

**Comment by tilgovi** [comment]
Template? No.

I think that whole function could move, though.

I don't like that we try to preserve cards across auth changes. It's a
performance hack, and it's a little risky (if something goes wrong we could
have non-public data still after sign-out).

Anyway, if the route reloads on auth change then it has to reconstruct its
view anyway. If annotations are cleaned up in the meantime, then everything
should work.

If this is too much work to rethink on this PR, we could merge this first,
but I'd like to see us at least think about what it would cost to improve.

File: h/static/scripts/stream-controller.coffee [hunk 3]
```diff
@@ -1,17 +1,70 @@
...
+    # Initialize cards
+    $scope.threadRoot = @threadRoot
+
+    $scope.lookup.hasAnnotation = (annotation) =>
+      @idTable[annotation.id]?.message
+
+    $scope.lookup.getAnnotationContainers = ->
+      $scope.threadRoot.children
+
+    $rootScope.$on 'beforeAnnotationCreated', (event, annotation) =>
```
Comment (tilgovi) [inline h/static/scripts/stream-controller.coffee]: These will need to be unbound on `$destroy`

**Comment by gergely-ujvari** [comment]
:+1: just for cleaning every template and don't try to be smart.
I'll check this.

**Comment by tilgovi** [comment]
The trickiest things to watch out for are that we don't lose unsaved highlights when signing in. Right now, those get instantly persisted.

**Comment by gergely-ujvari** [comment]
@tilgovi: I've created a branch to play here: https://github.com/hypothesis/h/tree/separate-controllers
It has no support, no fixed tests, but moved everything out from the AppController and it is much cleaner.
Do you support that way? (If yes, I'll make the rest)

**Comment by tilgovi** [comment]
I think I like it and I suspect it'll get simpler before we're done.

**Comment by gergely-ujvari** [comment]
Closing in favor of #2125

**Pull request closed without merging**